Added recipe API

Nephroid1 opened the pull request:
A proposal to the issue presented here: https://github.com/MinecraftForge/MinecraftForge/issues/1400

This API provides an abstraction for storing shaped and shapeless recipes for arbitrary machines.

### High Level Overview of this API:
Each crafting component is represented by an *AbstractIngredient* object. The *AbstractIngredient* has a *fulfilled()* method that, when given another *AbstractIngredient*, checks if it can be used as the crafting component. The *AbstractIngredient* also has a comparison method. The reason for this will be discussed with shapeless recipes.

A shaped recipe is represented by an array of *AbstractIngredient* objects. Each index in the array will correspond to a specific "slot" in a machine. A vanilla crafting table would have nine "slots," a vanilla furnace would have two "slots," and so on. These slots are implied when you create the shaped recipes for a machine.

The recipes for each machine is stored in a prefix tree in the *RecipeHandler* object for that machine. In order to check if a recipe exists, the *RecipeHandler* uses a filtered depth first search, using the *fulfilled()* method to eliminate recipes that don't match. To obtain a list of recipes for each result, a reverse lookup table is used instead.

The above representation of shaped recipes does not scale well for shapeless recipes. Each shapeless recipe must have a factorial number of recipe entries associated with it, which is too many. Thus shapeless recipes are stored separately from shaped recipes. With a well-ordering on the set of all recipe components of a machine, it is possible to sort the list of ingredients. Then two shapeless recipes are equal if and only if their sorted lists of ingredients are equal. With this in mind, shapeless recipes are stored in the same way as shaped recipes, with the exception that their lists of ingredients are sorted beforehand. Note that this ordering must function over all possible pairs of ingredients.

### Usage:
An example usage of this API is provided here: http://pastebin.com/gjAfSCTf

An example ordering is provided in the nested *TestIngredient* class. Note that this ordering must handle comparisons between different subclasses.

#### To register recipes:
1. For each machine you have, implement the methods in *IRecipeMachine*.
2. Register the machines using *RecipeManager*.*registerRecipeMachine()*.
3. Use *RecipeManager*.*addShapedRecipe()* and *RecipeManager*.*addShapelessRecipe()* to add recipes. Note that the *RecipeManager* keeps track of the last *IRecipeMachine* that you have interacted with.

To add secondary results, you can specify an optional *ArrayList* containing *ItemStack*-*Float* pairs. The float value represents the chance that the corresponding result is produced (if you want stochastic results), though you can use it for whatever you want.

#### To lookup recipes:
1. Get the *RecipeHandler* for your machine using *RecipeManager*.*getRecipeHandler()*.
2. The *RecipeHandler* provides a *lookupResult()* method which takes in an array of *AbstractIngredient*s.
3. The *RecipeHandler* also provides methods for reverse lookup.

diesieben07 commented:
Some thoughts:
In RecipeNode.fulfilled you check that the classes are equal. This should imho. check at least for subclasses. But it also does not allow for things like fluids where you want to treat the fluid container the same as the actual fluid.

Also you should not be using Hashtable, it is old and discouraged. Use HashMap instead (or, if you have a Map<T, List<V>> use Guava's Multimap.

Nephroid1 commented:
Diesieben07,

I'm not exactly sure what you mean about checking for subclasses. My reasoning was that if two AbstractIngredients are different classes, there's no way they could fulfill each other. However, I realize that this should be up to the user instead (i.e. in the fulfill method implemented by the user); that check was to prevent class cast exceptions, which should really be handled in the user implemented method instead. I have changed the code and the test/example to reflect this.

As for Hashtable vs Hashmap, I'm not too familiar with the differences between the two. I'll change it to Hashmap though, since that one seems more encouraged.

Thanks for the feedback.

diesieben07 commented:
What I meant that instead of checking for the exact same class (using `equals`) you should have used `instanceof` or rather one of the equivalent methods on the `Class` class.

The difference between `Hashtable` and `HashMap` is that Hashtable synchronizes on every method call, which may be desired, but may also not be desired. It is also the least effective way to achieve thread safety. If you need thread safety, `ConcurrentHashMap` should be used instead.

LexManos commented on src/main/java/net/minecraftforge/recipe/RecipeManager.java:
```diff
@@ -0,0 +1,74 @@
+        activeRecipeHandler = machineLookupTable.get(machine.getUniqueMachineName());
+    }
+
+    public static void addShapedRecipe(AbstractIngredient[] ingredientList, ItemStack result)
```
Everything below this line is useless.

LexManos commented on src/main/java/net/minecraftforge/recipe/RecipeManager.java:
```diff
@@ -0,0 +1,74 @@
+    {
+        if (machineLookupTable.containsKey(machine.getUniqueMachineName()))
+        {
+            System.out.format("[WARNING] Another machine with name %s has already been registered.\n Skipping this one.\n", machine.getUniqueMachineName());
```
Should give more information about this.. or just make getRecipeHandler() smarter and remove this function all together.

LexManos commented on src/main/java/net/minecraftforge/recipe/RecipeHandler.java:
```diff
@@ -0,0 +1,244 @@
+        private ArrayList<RecipeNode> possibleNextIngredients;
+        private AbstractIngredient ingredient;
+        private ItemStack result;
+        private ArrayList<Pair<ItemStack, Float>> stochasticResults;
```
Just a note, best practice is to keep this as loose as possible, so you'd want List all over the place not ArrayList.

LexManos commented on src/main/java/net/minecraftforge/recipe/RecipeHandler.java:
```diff
@@ -0,0 +1,244 @@
+        this.machineInstance = machineInstance;
+        shapedTrie = new RecipeNode(null);
+        shapelessTrie = new RecipeNode(null);
+        shapedReverseLookupTable = new HashMap<Item, ArrayList<Pair<AbstractIngredient[], Float>>>();
```
Probably shouldnt be hash maps, but more identity hash maps.

LexManos commented on src/main/java/net/minecraftforge/recipe/RecipeHandler.java:
```diff
@@ -0,0 +1,244 @@
+        if (shapedReverseLookupTable.get(i) != null)
+            return shapedReverseLookupTable.get(i);
+        
+        return new ArrayList<Pair<AbstractIngredient[], Float>>();
```
internal cache, or just return null from this.

LexManos commented:
General concept is fine, don't see it ever being used. It doesn't take into account anything such as container items, or modifications to things during the crafting.
Definitely not something to be used for Smelting, Aside from a abstraction out, I don't see any major benefits to this over the vanilla system.

Nephroid1 commented:
Hi LexManos,

Thanks for the replies. I've addressed most of the issues in the latest commit.

The intended use for this API isn't for replacing the vanilla system. It provides a standardized way of storing recipes so that every mod can access the recipe information for every machine in every other mod, without knowing what the other mods are or how those machines are implemented. As far as I know, there is no system in vanilla that lets you obtain the recipes for some arbitrary machine other mods - the only way to achieve what NEI does with modded machines is to have the other mods implement a plugin for NEI, which isn't feasible for most modders. As for how the stored recipes are used, that's up to the modders.

I'm not quite sure how this doesn't account for container items. If modders want to use a container item in a recipe, they can implement the fulfilled method in the abstract ingredient that represents that container item, which can be whatever they need for that container item. I don't see a case where it's not possible to represent a recipe using abstract ingredients.

I'm also not sure what you mean when you say modification to things during the crafting. This is just meant as a way to store recipes. It doesn't know anything about how those recipes are used.

Actuarius commented:
@williewillus added labels [Feature]
